src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,145 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import os
+from builtins import list, open, str
+from collections import defaultdict, namedtuple
+
+from future.moves.itertools import zip_longest
+from six import string_types
+
+from pants.base.build_environment import get_buildroot
+from pants.reporting.reporter import Reporter
+from pants.util.dirutil import safe_mkdir
+
+
+class JsonReporter(Reporter):
+  """Json reporting to files."""
+
+  _log_level_str = [
+    'FATAL',
+    'ERROR',
+    'WARN',
+    'INFO',
+    'DEBUG',
+  ]
+
+  # JSON reporting settings.
+  #   json_dir: Where the report files go.
+  Settings = namedtuple('Settings', Reporter.Settings._fields + ('json_dir',))
+
+  def __init__(self, run_tracker, settings):
+    super(JsonReporter, self).__init__(run_tracker, settings)
+    # The main report, and associated tool outputs, go under this dir.
+    self._json_dir = settings.json_dir
+
+    self._buildroot = get_buildroot()
+    self._report_file = None
+    self._output_files = defaultdict(dict)
+
+    # Results of the state of the build
+    self._results = {}
+    # Workunit stacks partitioned by thread
```

Comment:
Same nit.

Reply:
Also, the comment should mention what the keys are. Ideally, in fact, the name of the variable would. E.g., `self._thread_id_to_workunit_stack`. And the comment would explain *why* we have this data structure, e.g., "We use a stack to track the nested workunit traversal in each thread."

Reply:
Actually now that I read further, you're using the root workunit id as the key, not a thread id, so this should really be call `self._root_id_to_workunit_stack` or something. At this point the fact that the roots correspond to threads is neither here nor there. That is, nothing in your code cares about the concept of "threads" at all, i think?

Reply:
You're right in nothing here cares about threads - I'll make the update